dask/array/core.py

```diff
@@ -893,20 +897,72 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
         except AttributeError:
             dsk = {}
 
-        update = insert_to_ooc(src, tgt, lock=lock, region=reg)
-        keys.extend(update)
+            src = Array(
+                src.__dask_optimize__(
+                    src.__dask_graph__(),
+                    src.__dask_keys__()
+                ),
+                src.name,
+                src.chunks,
+                src.dtype,
+                src.shape
+            )
 
-        update.update(src.dask)
-        update.update(dsk)
-        updates.update(update)
+        each_store_dsk = insert_to_ooc(src, tgt, lock=lock, region=reg, keep=keep)
+
+        if keep:
+            load_names.append('load-%s' % src.name)
+            load_dsks.append(retrieve_from_ooc(each_store_dsk))
+
+        for each_store_key in each_store_dsk:
+            store_dlyds.append(
+                Delayed(
+                    each_store_key,
+                    sharedict.merge(
+                        (
+                            each_store_key,
+                            {each_store_key: each_store_dsk[each_store_key]}
+                        ),
+                        dsk,
+                        src.dask
+                    )
+                )
+            )
```

Comment:
It would also be fine here (and possibly more efficient) to just make a single `dsk` object for all of the delayed objects and differentiate them only by their key.  It is ok to have excess keys in a task graph.

Reply:
I'm assuming that the number of elements in each_store_dsk is proportional to the number of chunks.

Reply:
For the most part, we are not using the `Delayed` objects as can be seen in the code below. We actually unpack them in most cases and constructing some other object.

The one exception is `compute=True, keep=True` where we `persist` them. In that case, we'd still like to keep separate `Delayed` objects for each chunk as we'd like to be able to get some results from these Dask Arrays without having to block for the full result.

Reply:
We can still have separate delayed objects that share the same task graph.  Something like the following:

```python
dsk = {'a': (inc, 1), 'b': (dec, 1)}
a = Delayed('a', dsk)
b = Delayed('b', dsk)
```

In this case we constructed two delayed objects, but only constructed a single task graph.  I anticipate that generating millions of ShareDicts might be somewhat expensive (though we've tried hard to reduce these costs).  I would instead create a single task graph with all of the relevant tasks for all of the delayed objects and then just hand that single task graph to all of the delayed objects with distinct keys.

Reply:
Ah ok. Sorry, don't think I was following what you were suggesting before. This make sense.

Reply:
Have made some improvement on this by reusing the same task graph for each array being stored. As the task graph is unified and optimized over all `sources` earlier, this should pretty much address this issue.